lib/dockerregistry/inventory.go

```diff
@@ -51,59 +54,373 @@ func getSrcRegistry(rcs []RegistryContext) (*RegistryContext, error) {
+	srcRegsSeen := make(map[RegistryName]string)
+	for _, mfest := range mfests {
+		if mfestFilepath, seen := srcRegsSeen[mfest.srcRegistry.Name]; seen {
+			// nolint[lll]
+			return fmt.Errorf("source registry '%s' defined in multiple manifests:\n- '%s'\n- '%s'\n", mfest.srcRegistry.Name, mfestFilepath, mfest.filepath)
+		}
+		srcRegsSeen[mfest.srcRegistry.Name] = mfest.filepath
+	}
+
+	// If two manifests are renaming images, then they should not share any
+	// rename information (should be mutually exclusive). We use a separate loop
+	// for clarity.
+	renamesSeen := make(map[RegistryImagePath]string)
+	for _, mfest := range mfests {
+		for _, rename := range mfest.Renames {
+			for _, regImgPath := range rename {
+				if mfestFilepath, seen := renamesSeen[regImgPath]; seen {
+					// nolint[lll]
+					return fmt.Errorf("rename key '%s' found in multiple manifests:\n- '%s'\n- '%s'\n", regImgPath, mfestFilepath, mfest.filepath)
+				}
+				renamesSeen[regImgPath] = mfest.filepath
+			}
+		}
+	}
+
+	// Force all manifests to share the EXACT same non-src-registry
+	// configuration. This is because they should all be promoting to the same
+	// central registry, with the same creds.
+	canonicalDstRegisries := []RegistryContext{}
+	for _, registry := range mfests[0].Registries {
+		if !registry.Src {
+			canonicalDstRegisries = append(canonicalDstRegisries, registry)
+		}
+	}
+	for _, mfest := range mfests[1:] {
+		dstRegistries := []RegistryContext{}
+		for _, registry := range mfest.Registries {
+			if !registry.Src {
+				dstRegistries = append(dstRegistries, registry)
+			}
+		}
+
+		if !reflect.DeepEqual(dstRegistries, canonicalDstRegisries) {
+			return fmt.Errorf("different destination registries found in manifests:\n- '%s'\n- '%s'\n", mfests[0].filepath, mfest.filepath)
+		}
+	}
+
+	return nil
+}
+
+// ToPromotionEdges converts a list of manifests to a set of edges we want to
+// try promoting.
+func ToPromotionEdges(mfests []Manifest) map[PromotionEdge]interface{} {
+	edges := make(map[PromotionEdge]interface{})
+	// nolint[lll]
+	for _, mfest := range mfests {
+		for _, image := range mfest.Images {
+			for digest, tagArray := range image.Dmap {
+				for _, tag := range tagArray {
+					for _, rc := range mfest.Registries {
+						if rc == *mfest.srcRegistry {
+							continue
+						}
+						edge := PromotionEdge{
+							SrcRegistry: *mfest.srcRegistry,
+							SrcImageTag: ImageTag{
+								ImageName: image.ImageName,
+								Tag:       tag},
+							Digest:      digest,
+							DstRegistry: rc,
+						}
+
+						// Renames change how edges are created.
+						regImgPath := RegistryImagePath(mfest.srcRegistry.Name) + "/" + RegistryImagePath(image.ImageName)
+						if renames, ok := mfest.renamesDenormalized[regImgPath]; ok {
+							if imgName, ok := renames[rc.Name]; ok {
+								edge.DstImageTag = ImageTag{
+									ImageName: imgName,
+									Tag:       tag}
+								edges[edge] = nil
```

Comment:
Use of `nil` as the key is pretty confusing.  I usually use `map[foo]bool` or if I'm feeling particularly efficient, `map[foo]struct{}`

Reply:
Wait, `edge` is the key. The value is `nil`. This is using `map[foo]interface{}`, which I suppose is close enough to `map[foo]struct{}`.

Reply:
Right, seeing you setting the values to nil, I was assuming you were zeroing them out to remove them.  Bad assumption, but it's not the common pattern...